src/hotspot/cpu/s390/stubGenerator_s390.cpp

```diff
@@ -1469,10 +1469,127 @@ class StubGenerator: public StubCodeGenerator {
   }
 
 
+  // Helper for generate_unsafe_setmemory
+  //
+  // Atomically fill an array of memory using 1-, 2-, 4-, or 8-byte chunks and return.
+  static void do_setmemory_atomic_loop(int elem_size, Register dest, Register size, Register byteVal,
+                                       MacroAssembler *_masm) {
+
+    NearLabel L_Loop, L_Tail; // 2x unrolled loop
+    Register tmp = Z_R1; // R1 is free at this point
+
+    if (elem_size > 1) {
+      __ rotate_then_insert(byteVal, byteVal, 64 - 2 * 8 , 63 - 8,  8, 0);
+    }
+
+    if (elem_size > 2) {
+      __ rotate_then_insert(byteVal, byteVal, 64 - 2 * 16, 63 - 16, 16, 0);
+    }
+
+    if (elem_size > 4) {
+      __ rotate_then_insert(byteVal, byteVal, 64 - 2 * 32, 63 - 32, 32, 0);
+    }
+
+    __ z_risbg(tmp, size, 32, 128/* risbgz */ + 63, 64 - exact_log2(2 * elem_size), 0); // just do the right shift and set cc
+    __ z_bre(L_Tail);
+
+    __ align(16); // loop alignment
+    __ bind(L_Loop);
+    __ store_sized_value(byteVal, Address(dest, 0), elem_size);
+    __ store_sized_value(byteVal, Address(dest, elem_size), elem_size);
+    __ z_agfi(dest, 2 * elem_size);
+    __ z_brct(tmp, L_Loop);
+
+    __ bind(L_Tail);
+    __ z_nilf(size, elem_size);
+    __ z_bcr(Assembler::bcondEqual, Z_R14);
+    __ store_sized_value(byteVal, Address(dest, 0), elem_size);
+    __ z_br(Z_R14);
+  }
+
+  //
+  //  Generate 'unsafe' set memory stub
+  //  Though just as safe as the other stubs, it takes an unscaled
+  //  size_t (# bytes) argument instead of an element count.
+  //
+  //  Input:
+  //    Z_ARG1   - destination array address
+  //    Z_ARG2   - byte count (size_t)
+  //    Z_ARG3   - byte value
+  //
+  address generate_unsafe_setmemory(address unsafe_byte_fill) {
+    __ align(CodeEntryAlignment);
+    StubCodeMark mark(this, StubGenStubId::unsafe_setmemory_id);
+    unsigned int start_off = __ offset();
+
+    // bump this on entry, not on exit:
+    // inc_counter_np(SharedRuntime::_unsafe_set_memory_ctr);
+
+    {
+      NearLabel L_fill8Bytes, L_fill4Bytes, L_fillBytes, L_exit;
+
+      const Register dest = Z_ARG1;
+      const Register size = Z_ARG2;
+      const Register byteVal = Z_ARG3;
+      const Register rScratch1 = Z_R1_scratch;
+      // fill_to_memory_atomic(unsigned char*, unsigned long, unsigned char)
+
+      // Check for pointer & size alignment
+      __ z_ogrk(rScratch1, dest, size);
+
+      __ z_nill(rScratch1, 7);
+      __ z_bre(L_fill8Bytes); // branch if 0
+
+
+      __ z_nill(rScratch1, 3);
+      __ z_bre(L_fill4Bytes); // branch if 0
+
+      __ z_nill(rScratch1, 1);
+      __ z_brne(L_fillBytes); // branch if not 0
+
+      // Mark remaining code as such which performs Unsafe accesses.
+      UnsafeMemoryAccessMark umam(this, true, false);
+
+      // At this point, we know the lower bit of size is zero and a
+      // multiple of 2
+      do_setmemory_atomic_loop(2, dest, size, byteVal, _masm);
+
+      __ align(16);
```

Comment:
What is this alignment good for?

Reply:
Branch target alignment. There is no fallthrough path from before this point. Should it be 32?